### DISCUSSION TITLE
Nextcloud App

### DISCUSSION THREAD

**Discussion by christianlupus** [iSoron/uhabits#1099] [Feature Requests]
I have the idea to allow syncing the state/habit data to/via a dedicated nextcloud app. This would tackle #702 and potentially touch #486.

As I am a contributor to another nextcloud app, I would be willing to help on the NC side of the feature request. For me, a web version (eventually with a mail reminder) would be a hard requirement for the usage of the app. Therefore I was thinking of solutions.

Before I start investing time, I wanted to make sure, this would be considered a good idea and supported by the main developers here (most probably @iSoron).

**Comment by iSoron** [comment]
Thank you for the suggestion, @christianlupus, but it looks to me like NextCloud integration would require us to rewrite the entire app in PHP, which is not feasible. It's also unclear to me how would this solution benefit current users of the Android app.

**Reply by christianlupus** [reply]
@iSoron, I am not thinking of rewiring the complete app in PHP. This would throw away any so far made progress. No worries, I am not dumb.

I am thinking of a (REST-based) API interface that allows synchronizing the saved habits plus the achievements between the app (as it is currently plus a few extensions) and an NC app, that will store this information per user on the server. This would allow to

1. Store and save the data in the NC database thus adding a backup layer in case of phone outage
2. Eventually modify/track habits/achievements in the NC web interface. Don't worry about the implementation here.
3. Central storage would allow sharing the successes between multiple devices.

What would be required to get this running?

- A clear definition of what information needs to be stored
- A documentation of how to get the sync issue addressed (how to guarantee consistency etc)
- Definition of an appropriate API for the NC
- Implementing the usage of the API plus its configuration (which server to use, which credentials etc) in the habit tracker
- Implementation of the NC side in a new app that needs to be created. Here I can help/come into play.

I suspect that the first few points can be done by looking at the current code of the app, although this might be taking some time. The definition of the API should be done in close cooperation to avoid confusion. I could start with the NC side but this would involve quite some guessing and I am not capable to help with the Kotlin code. So, for this to work, I needed your help on the Kotlin side. Therefore, I asked if this extension towards NC would be appreciated or discarded.
In the latter case, I would save the time required for the PHP/JS/... development for my other tasks.

Does this make now more sense to you?

**Comment by madpipeline** [comment]
I am very interested in this idea as well. But for this to happen, even just the basic functionality of using NC as a storage the app should be able to not only routinely export and upload it's data, but at some point also handle conflicts between local and remote data. Multi device support is not possible without automatic conflict resolution. Most likely implemented on the clients.

**Reply by christianlupus** [reply]
This is nice to hear.
Regarding the conflict resolution issue, this is one of the things that need discussion. This might be partially solved on the server: When there is a web frontend within the NC app, the solution is needed there as well. I was considering splitting this up into two parts: Automatic resolution of mergeable conflicts in the server and manual interaction in case of real conflicts. But as I told you, I am not yet fixed and this needs further discussion.

**Reply by madpipeline** [reply]
Personally I'd rather have the server behave just like a remote storage solution and have the clients have the conflict resolution logic. This way the app won't get to be dependent on the server, and will function the same as stand-alone.

Regarding the CR logic, The Joplin app is a good example. We can just timestamp every change, and check the timestamps to see which is the latest change, and if the changes can be merged or user interaction is needed to solve the conflict.


Either way,  for starters just an automatic export to a remote location in NC is good enough for a first version.

**Reply by christianlupus** [reply]
> Personally I'd rather have the server behave just like a remote storage solution and have the clients have the conflict resolution logic. This way the app won't get to be dependent on the server, and will function the same as stand-alone.

That is right. I am just no fan of code duplication (client on android + web client). But the other way (client resolves the conflicts) makes the server part simpler.

> Regarding the CR logic, The Joplin app is a good example. We can just timestamp every change, and check the timestamps to see which is the latest change, and if the changes can be merged or user interaction is needed to solve the conflict.

Funny, you mention Joplin here. I use it but never elaborated their CR logic. Do you know of (good) documentation or do you have direct knowledge?

The reason I ask: I can imagine having each version attached with a time stamp. The question is now, if a conflict arises, if the changes made are compatible (e.g. different habits/files/....), or a real conflict where a human must resolve it. I am merely thinking similarly to git, where you have two heads to merge and a common ancestor. You create the patches from the ancestor to both ends and check of the patches can be applied together without issues or if these are contradicting. That would mean we had to save all historic versions on the server (which should be doable).
If you had only the very last version present, I consider it harder to distinguish the contradicting cases from the others. Consider on two unsynced devices a habit is marked as done for the week. Then the devices resync one by one. The first is no problem updating the server timestamp. The second will only know that a new version was published.
I guess, if you duplicated the latest synced state, you could calculate the delta between this latest sync and the current state on the client and compare in the same manner the server version with the saved root. I am not 100% sure on this and if you @madpipeline are referring to such a structure.

> Either way, for starters just an automatic export to a remote location in NC is good enough for a first version.

Just an export will not help much (at least in my case). I want to sync over multiple devices and ideally have a web interface to have the habits ready on the laptop instead of just the phone.

I want to train myself e.g. to daily have a look in the task list (management of todos in Redmine) if something is due.

**Reply by iSoron** [reply]
@christianlupus @madpipeline We already have a working solution for all the sync issues you are discussing. It has already been implemented and it will become available in Loop 2.1. If you enable device sync, Loop will periodically encrypt your database locally, using a strong cryptographic key kept only in your device, then upload the encrypted database to a centralized server. You will also be able to run the server yourself (self-host) if you don't want to use the public instance.

Assuming that this is already deployed, I can see two additional feature requests related to Nextcloud:

1. **I would like to self-host and I already run Nextcloud. Could Loop upload its data there, instead?** I think it's possible. I will personally not spend time implementing this, since I think only a small fraction of our users run Nextcloud, but if you could implement a server within Nextcloud that follows the same API as our dedicated server, everything should work automatically. Note that the server API is not documented at this time and it's still subject to change.
2. **Since the data is already uploaded to Nextcloud, could I manage my habits there too?** This would require either reimplementing the app in PHP or finishing #486.

**Reply by christianlupus** [reply]
@iSoron that is good to hear. Is it possible to share the habits on multiple devices then? I understood the statement with the crypto in a way that the saved data is only readable from that single device.

Do you have a plan on documenting the API in the (near) future? As I wrote, I am currently more interested in a web-based frontend (usage at laptop) ideally with mail reminders. See also #486, #483, and #245.

**Reply by madpipeline** [reply]
It's good to hear that there is already work being done on the sync and server. I don't have strong feeling towards hosting it in NC. I'm ok with stand-alone. It would be good if the server app is delivered as a Docker image, for quick deployments.

If the server also has a front-end, so we can use/manage the habits it would be great. At this moment, for me at least it's not a necessity.

@iSoron I have Go, python, Docker and others in my knowledge bag. Let me know if I can help in any way. I specialize in process management and CI/CD patterns.

**Reply by iSoron** [reply]
@christianlupus It will be possible to have the same set of habits in multiple devices. We can make the server API published after it's stable.

@madpipeline Thank you for volunteering. After we have the first prototype working, I will create specific issues on possible improvements or extensions. Any help would be greatly appreciated.

**Reply by christianlupus** [reply]
> @christianlupus It will be possible to have the same set of habits in multiple devices. We can make the server API published after it's stable.

@iSoron That is good to hear. If I can help with something, ping me.

**Comment by gynnantonix** [comment]
I'd be happy to help with making this a reality. I don't know PHP or Kotlin, though.
